File: velocity/src/main/java/ru/privatenull/pnauth/velocity/dialog/VelocityDialogServiceFactory.java
package ru.privatenull.pnauth.velocity.dialog;

import com.velocitypowered.api.proxy.Player;
import com.velocitypowered.api.proxy.ProxyServer;
import org.slf4j.Logger;

/** @deprecated Legacy factory retained for rollback reference. */
@Deprecated(forRemoval = false)
final class VelocityDialogServiceFactory {
    private VelocityDialogServiceFactory() {
    }

    static VelocityDialogService create(ProxyServer proxy, Logger logger,
                                        VelocityDialogService.SubmissionHandler submissions) {
        try {
            Class<?> type = Class.forName(
                    "ru.privatenull.pnauth.velocity.dialog.PacketEventsVelocityDialogService",
                    true, VelocityDialogServiceFactory.class.getClassLoader());
            return (VelocityDialogService) type
                    .getConstructor(VelocityDialogService.SubmissionHandler.class)
                    .newInstance(submissions);
        } catch (ReflectiveOperationException | LinkageError exception) {
            logger.warn("Bundled PacketEvents dialog integration failed; using commands.", exception);
            return UnavailableDialogService.INSTANCE;
        }
    }

    private enum UnavailableDialogService implements VelocityDialogService {
        INSTANCE;
        public boolean available() { return false; }
        public void show(Player player, DialogForm form) { }
        public void clear(Player player) { }
    }
}

File: velocity/src/main/kotlin/ru/privatenull/pnauth/velocity/dialog/VelocityDialogServiceFactory.kt
package ru.privatenull.pnauth.velocity.dialog

import com.velocitypowered.api.proxy.Player
import com.velocitypowered.api.proxy.ProxyServer
import org.slf4j.Logger

internal object VelocityDialogServiceFactory {

    @JvmStatic
    fun create(
        proxy: ProxyServer,
        logger: Logger,
        submissions: VelocityDialogService.SubmissionHandler
    ): VelocityDialogService {
        return try {
            val type = Class.forName(
                "ru.privatenull.pnauth.velocity.dialog.PacketEventsVelocityDialogService",
                true, VelocityDialogServiceFactory::class.java.classLoader
            )
            type.getConstructor(VelocityDialogService.SubmissionHandler::class.java)
                .newInstance(submissions) as VelocityDialogService
        } catch (exception: Throwable) {
            if (exception is ReflectiveOperationException || exception is LinkageError) {
                logger.warn("Bundled PacketEvents dialog integration failed; using commands.", exception)
                UnavailableDialogService.INSTANCE
            } else {
                throw exception
            }
        }
    }

    private enum class UnavailableDialogService : VelocityDialogService {
        INSTANCE;

        override fun available(): Boolean = false
        override fun show(player: Player, form: VelocityDialogService.DialogForm) {}
        override fun clear(player: Player) {}
    }
}
